### PR TITLE
Allow admins to change roles ctd

### LINKED ISSUE #1271: Not able to upgrade users
Thanks for creating an issue! Please fill out this form so we can be sure to have all the information we need, and to minimize back and forth.

* **What are we trying to do?**
Bump new users in BRAAF to admin. Currently new users are created as CM and a change to any other class of user doesn't save.

* **What feature or behavior is this required for?**
Fund's ability to manage users.

* **How could we solve this issue? (Not knowing is okay!)**
Make it so any admin can bump users to another class.

* **Anything else?**
Our engineering team rocks!

### PR DESCRIPTION
**Pull request by colinxfleming**
I rule and have completed some work on Case Manager that's ready for review!

Continues #1272 - Made a quick change to allow roles to be permitted on update.

This pull request makes the following changes:
* Whitelists user role so an admin can change it
* Adds a safety so admins cannot demote themselves
* Overhauls the users controller test

It relates to the following issue #s: 
* Fixes #1271

### CODE REVIEW

**Comment by colinxfleming** [comment]
I think I'm going to use this as an opportunity to overhaul the users controller tests. Right now they're sort of all over the place (my fault!) and don't have fantastic coverage.

**Comment by colinxfleming** [comment]
I'm going to file a ticket to split up the call list management functionality and the admin user management functionality. The users controller is kinda heavy as it is, the call list stuff needs a refactor, and the call list stuff doesn't need to inherit a bunch of confusing junk from devise.

@tingaloo can I get a review? Sorry this is so heavy.

**Comment by colinxfleming** [comment]
We have a brakeman failure on role.

```
Confidence: Medium
Category: Mass Assignment
Check: PermitAttributes
Message: Potentially dangerous key allowed for mass assignment
Code: params.require(:user).permit(:name, :email, :role)
File: app/controllers/users_controller.rb
Line: 117
```

**Comment by colinxfleming** [comment]
Potential fix is making 'set role' to be a button with its own endpoint. Not restful but likely won't trigger the warning.

**Comment by colinxfleming** [comment]
Hot damn did this ever go off the rails. This moves the endpoints to change roles out of update, throws them behind an auth wall, adds buttons on the user edit form, and throws in a systemtest for good measure. I think that'll squelch the brakeman warning, so after that it'll be whatever else.

**Comment by colinxfleming** [comment]
We're getting incidental failures, so I think this is ready for Serious Review.

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -29,6 +29,10 @@ def confirm_admin_user
     redirect_to root_url unless current_user.admin?
   end
 
+  def confirm_admin_user_async
```
Comment (tingaloo) [inline app/controllers/application_controller.rb]: Not sure if ```async``` is necessary.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -121,4 +141,14 @@ def retrieve_patients
     @patient = Patient.find params[:id]
     @urgent_patient = Patient.where(urgent_flag: true)
   end
+
+  def user_not_demoting_themself(user)
```
Comment (tingaloo) [inline app/controllers/users_controller.rb]: Can add ```?``` at end of method name

File: app/helpers/users_helper.rb [hunk 3]
```diff
@@ -15,4 +15,11 @@ def user_lock_status(user)
       'Active'
     end
   end
+
+  def role_toggle_button(user, new_role)
```
Comment (tingaloo) [inline app/helpers/users_helper.rb]: This is pretty cool, does this dynamically creates the button text based on roles?

File: app/models/concerns/user_searchable.rb [hunk 4]
```diff
@@ -6,13 +6,12 @@ module UserSearchable
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
 
   class_methods do
-
     def search(name_or_email_str)
-      name_regexp = /#{Regexp.escape(name_or_email_str)}/i
-      email_regexp = (VALID_EMAIL_REGEX).match(name_or_email_str)
+      regexp = /#{Regexp.escape(name_or_email_str)}/i
+      # email_regexp = (VALID_EMAIL_REGEX).match(name_or_email_str)
```
Comment (tingaloo) [inline app/models/concerns/user_searchable.rb]: can remove this?

File: test/controllers/users_controller_test.rb [hunk 5]
```diff
@@ -2,49 +2,180 @@
...
+      end
+    end
+
+    it 'should show errors if creation fails' do
+      attributes = attributes_for(:user)
+      attributes[:name] = nil
+      assert_no_difference 'User.count' do
+        post users_path, params: { user: attributes }
+      end
+      assert_includes response.body, 'can&#39;t be blank'
```
Comment (tingaloo) [inline test/controllers/users_controller_test.rb]: What an odd assertion

Comment (colinxfleming) [reply test/controllers/users_controller_test.rb]: haha. it escapes the apostrophe like a real jerk.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (colinxfleming) [reply app/controllers/application_controller.rb]: We already have a `confirm_admin_user` synchronous that redirects to root, sad to say.

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (colinxfleming) [reply app/controllers/users_controller.rb]: Good call! 61587d11

File: app/helpers/users_helper.rb (hunk 3 above)
Comment (colinxfleming) [reply app/helpers/users_helper.rb]: yep! Spits out a button that, when you click it, patches to those endpoints.

File: app/models/concerns/user_searchable.rb (hunk 4 above)
Comment (colinxfleming) [reply app/models/concerns/user_searchable.rb]: Yep! 1013abd0

File: test/controllers/users_controller_test.rb [hunk 6]
```diff
@@ -2,49 +2,180 @@
...
   end
 
-  describe 'create method' do
-    it 'should raise if user is not admin' do
+  describe 'authentication for restricted endpoints' do
+    %w[users_path new_user_path].each do |endpoint|
+      it "should redirect if user is not admin - #{endpoint}" do
+        User.role.values.reject { |role| role == :admin }.each do |role|
+          @user.update role: role
+          get send(endpoint.to_sym)
```
Comment (tingaloo) [inline test/controllers/users_controller_test.rb]: I've never seen send() used before, is it being used as a rest call here?

Comment (Kevin-Wei) [reply test/controllers/users_controller_test.rb]: Send is being used here to call methods :)

**Comment by colinxfleming** [comment]
There's an alarming set of flaky tests here but YOLO. Thanks @tingaloo and @Kevin-Wei !

**Pull request merged**